id/flexid.go

```diff
@@ -0,0 +1,239 @@
+package id
+
+import (
+	"fmt"
+	"regexp"
+	"strings"
+
+	"github.com/manifoldco/go-manifold"
+	"github.com/manifoldco/go-manifold/errors"
+)
+
+const (
+	pathSeperator string = `\`
+
+	// ManifoldDomain is the domain name used to identify Manifold IDs
+	ManifoldDomain Domain = "manifold.co"
+)
+
+var (
+	// domainRegex expects that the string is a valid and easy to understand hostname
+	domainRegex = regexp.
+			MustCompile(`^([a-zA-Z0-9-_]+\.)*[a-zA-Z0-9][a-zA-Z0-9-_]+\.[a-zA-Z]{2,11}?$`)
+	// idRegex expects that an ID at least has a length of one, an only includes
+	//  characters expected in Base64 encoded values, GUIDs and UUIDs
+	idRegex = regexp.MustCompile(`^\{?[a-zA-Z0-9+/-_]{1,256}={0,2}\}?$`)
+
+	errInvalidParts = manifold.NewError(errors.BadRequestError,
+		"Invalid CompositeID, expected 3 parts, Domain, Type, and ID")
+	errInvalidDomain = manifold.NewError(errors.BadRequestError,
+		"Invalid CompositeID, expected a valid Domain in the first segment")
+	errInvalidType = manifold.NewError(errors.BadRequestError,
+		"Invalid CompositeID, expected a valid Type in the last segment")
+	errInvalidID = manifold.NewError(errors.BadRequestError,
+		"Invalid CompositeID, expected a valid ID in the last segment")
+
+	// ErrNotAManifoldID is an error returned when a CompositeID is expected to
+	//  be a ManifoldID, but is not.
+	ErrNotAManifoldID = manifold.NewError(errors.BadRequestError,
+		"Malformed ManifoldID, expected form `manifold.co/TYPE/MANIFOLDID`")
+	// ErrManifoldIDTypeMismatch is an error returned when a CompositeID is expected to
+	//  be a ManifoldID, but is not because the type does not match.
+	ErrManifoldIDTypeMismatch = manifold.NewError(errors.BadRequestError,
+		"Invalid ManifoldID, expected TYPE from `manifold.co/TYPE/ID` to match ID Type")
+)
+
+// Domain is a string that can be Validated based on Regex to expect a string
+//  that represents a Domain
+type Domain string
+
+// Validate ensures the name value is valid
+func (d Domain) Validate(_ interface{}) error {
+	if domainRegex.Match([]byte(d)) {
+		return nil
+	}
+
+	return errInvalidDomain
+}
+
+// ExternalID is a string that can be Validated based on Regex to expect a string
+//  representative of an ExternalID
+type ExternalID string
+
+// Validate ensures the name value is valid
+func (eid ExternalID) Validate(_ interface{}) error {
+	if idRegex.Match([]byte(eid)) {
+		return nil
+	}
+
+	return errInvalidID
+}
+
+// CompositeID is an ID that also includes the domain, and type of the identifier.
+//  Composed as: DOMAIN / TYPE / ID
+//  Example: manifold.co/user/2003btphq7z6dzvjut370jkvkdgcp
+//  Has `manifold.co` as the domain, a type of `user`, followed by the Manifold ID.
+type CompositeID interface {
+	Domain() Domain
+	Type() manifold.Label
+	ID() ExternalID
+	AsFlexID() *FlexID
+	// Common interfaces
+	String() string
+	Validate(interface{}) error
```

Comment:
Do we have to have the interface here? Are we going to use it at one point?

Reply:
This is from OpenAPI, this is what it specifies. I'll add a comment to explain.